packages/contracts-bedrock/src/cannon/libraries/MIPS64Instructions.sol

```diff
@@ -0,0 +1,920 @@
+                    return signExtend(uint64(uint32(_rs) + uint32(_rt)), 32);
+                }
+                // sub
+                else if (_fun == 0x22) {
+                    return signExtend(uint64(uint32(_rs) - uint32(_rt)), 32);
+                }
+                // subu
+                else if (_fun == 0x23) {
+                    return signExtend(uint64(uint32(_rs) - uint32(_rt)), 32);
+                }
+                // and
+                else if (_fun == 0x24) {
+                    return (_rs & _rt);
+                }
+                // or
+                else if (_fun == 0x25) {
+                    return (_rs | _rt);
+                }
+                // xor
+                else if (_fun == 0x26) {
+                    return (_rs ^ _rt);
+                }
+                // nor
+                else if (_fun == 0x27) {
+                    return ~(_rs | _rt);
+                }
+                // slti
+                else if (_fun == 0x2a) {
+                    return int64(_rs) < int64(_rt) ? 1 : 0;
+                }
+                // sltiu
+                else if (_fun == 0x2b) {
+                    return _rs < _rt ? 1 : 0;
+                }
+                // dadd
+                else if (_fun == 0x2c) {
+                    return (_rs + _rt);
+                }
+                // daddu
+                else if (_fun == 0x2d) {
+                    return (_rs + _rt);
+                }
+                // dsub
+                else if (_fun == 0x2e) {
+                    return (_rs - _rt);
+                }
+                // dsubu
+                else if (_fun == 0x2f) {
+                    return (_rs - _rt);
+                }
+                // dsll
+                else if (_fun == 0x38) {
+                    return _rt << ((_insn >> 6) & 0x1f);
+                }
+                // dsrl
+                else if (_fun == 0x3A) {
+                    return _rt >> ((_insn >> 6) & 0x1f);
+                }
+                // dsra
+                else if (_fun == 0x3B) {
+                    return uint64(int64(_rt) >> ((_insn >> 6) & 0x1f));
+                }
+                // dsll32
+                else if (_fun == 0x3c) {
+                    return _rt << (((_insn >> 6) & 0x1f) + 32);
+                }
+                // dsrl32
+                else if (_fun == 0x3e) {
+                    return _rt >> (((_insn >> 6) & 0x1f) + 32);
+                }
+                // dsra32
+                else if (_fun == 0x3f) {
+                    return uint64(int64(_rt) >> (((_insn >> 6) & 0x1f) + 32));
+                } else {
+                    revert("MIPS64: invalid instruction");
+                }
+            } else {
+                // SPECIAL2
+                if (_opcode == 0x1C) {
+                    // mul
+                    if (_fun == 0x2) {
+                        return signExtend(uint64(uint32(_rs) * uint32(_rt)), 32);
```

Comment:
`mul` is an interesting one. This instruction isn't part of MIPS release 1. So this is dead code. I'll add more testing nonetheless.

Reply:
This turned out to be critical. Good callout.

Reply:
I added a test that fuzzes various inputs for `mul/mult/multu`. The expected output isn't checked for correctness, but instead checks for consistency between the Go and solidity VMs.  
This test now runs in CI. It would be worthwhile to expand this to other instructions later.